File: Python/Minimum Number of Vertices to Reach All Nodes.py
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(n)
    """

    def findSmallestSetOfVertices(self, n: int, edges: List[List[int]]) -> List[int]:
        seen = [False] * n

        for _, to in edges:
            seen[to] = True

        return [v for v, is_seen in enumerate(seen) if not is_seen]

File: Java/Minimum Number of Vertices to Reach All Nodes.java
import java.util.ArrayList;
import java.util.List;

public class Solution {

    /**
     * Time:   O(n)
     * Memory: O(n)
     */
    public List<Integer> findSmallestSetOfVertices(int n, List<List<Integer>> edges) {
        List<Integer> res = new ArrayList<>();
        int[] seen = new int[n];

        for (List<Integer> e : edges)
            seen[e.get(1)] = 1;

        for (int i = 0; i < n; ++i)
            if (seen[i] == 0)
                res.add(i);

        return res;
    }
}
